### PR TITLE
Reverse idempotency in S3Parq

### PR DESCRIPTION
**Pull request by natie-ichain**
Right now, s3parq returns an empty dataset when Reverse=True and only the comparison dataset contains data. This is unintentional and breaks the idempotency of DatasetDiff. We need to fix the idempotency of s3parq to properly handle repeated pipeline runs in Airflow.

Definition of Done: 
s3parq fixed to expected values
s3parq unit tests all pass

### CODE REVIEW

File: tests/test_integration.py [hunk 1]
```diff
@@ -5,7 +5,7 @@
 import dfmock
 from s3parq.testing_helper import df_equal_by_set
 from s3parq.publish_parq import publish
-from s3parq.fetch_parq import fetch
+from s3parq.fetch_parq import fetch, get_all_partition_values
```
Comment (schafrn) [inline tests/test_integration.py]: Does this get used anywhere in here?

File: s3parq/fetch_parq.py [hunk 2]
```diff
@@ -329,21 +332,26 @@ def _get_filtered_key_list(typed_parts: dict, filters, key) -> List[str]:
...
-                    )
-
-    def construct_paths(typed_parts, previous_fil_keys: List[str]) -> None:
-        if len(typed_parts) > 0:
-            part = typed_parts.popitem(last=False)
+        matched_part_vals.clear()
+        fil = next((f for f in filters if f['partition'] == part), False)
+        if fil:
+            comparison = OPS[fil['comparison']]
+            for v in fil['values']:
```
Comment (jacobtobias) [inline s3parq/fetch_parq.py:344]: is this always going to have a `values`?  if we can't promise that and we don't want it to fail if it's missing values, we can use `for v in fil.get('values', [])`.  the `get` will default to an empty array if `values` is missing.

if when `values` is missing something is clearly not right and we'd prefer for it to fail with a `KeyError`, i think this is ok

File: tests/test_integration.py (hunk 1 above)
Comment (schafrn) [reply tests/test_integration.py]: Cool, thanks for fixing!

File: s3parq/fetch_parq.py [hunk 3]
```diff
@@ -329,21 +332,26 @@ def _get_filtered_key_list(typed_parts: dict, filters, key) -> List[str]:
...
+                for x in part_values:
+                    if comparison(x, v):
+                        matched_part_vals.add(x)
+            matched_parts[part] = matched_part_vals.copy()
+        else:
+            matched_parts[part] = part_values
+            
+    def construct_paths(matched_parts, previous_fil_keys: List[str]) -> None:
```
Comment (jacobtobias) [inline s3parq/fetch_parq.py:352]: What `type` is `matched_parts` supposed to be?  if we know `previous_fil_keys` is expected to be a `List[str]`, we should try to be consistent and be explicit about what type this variable will be

File: s3parq/fetch_parq.py (hunk 2 above)
Comment (natie-ichain) [reply s3parq/fetch_parq.py:344]: Each filter requires a partition, a series of values, and a comparison operator. This is the defined schema for filters called through s3parq. If a partition is found for a particular filter, I think it is a fair assumption for the filter to contain a list of values and an operator. The original code also reflected this assumption.

File: s3parq/fetch_parq.py (hunk 3 above)
Comment (natie-ichain) [reply s3parq/fetch_parq.py:352]: matched_parts is defined in `_get_filtered_key_list()` to be an OrderedDict, similar to typed_parts (at the top of the function). While typed_parts contains _all_ the partition values for a particular dataset, matched_parts removes partition values based on the filters passed to the function.

**Review by schafrn — APPROVED** [review]
👍 Thanks for the fix!

**Pull request merged**